src/emulator/storage/apis/gcloud.ts

```diff
@@ -261,18 +259,65 @@ export function createCloudEndpoints(emulator: StorageEmulator): Router {
     const md = storageLayer.getMetadata(req.params.bucketId, req.params.objectId);
 
     if (!md) {
-      res.sendStatus(404);
-      return;
+      return sendObjectNotFound(req, res);
     }
 
     return sendFileBytes(md, storageLayer, req, res);
   });
 
+  gcloudStorageAPI.post(
+    "/b/:bucketId/o/:objectId/:method(rewriteTo|copyTo)/b/:destBucketId/o/:destObjectId",
+    (req, res, next) => {
+      const md = storageLayer.getMetadata(req.params.bucketId, req.params.objectId);
+
+      if (!md) {
+        return sendObjectNotFound(req, res);
+      }
+
+      if (req.params.method === "rewriteTo" && req.query.rewriteToken) {
+        // Don't yet support multi-request copying
+        return next();
+      }
+
+      const metadata = storageLayer.copyFile(
+        req.params.bucketId,
+        req.params.objectId,
+        req.params.destBucketId,
+        req.params.destObjectId,
+        req.body
+      );
+
+      if (!metadata) {
+        res.sendStatus(400);
```

Comment:
Because [`StorageLayer#copyFile`](https://github.com/bookcreator/firebase-tools/blob/0191163aa697f2f7cc94c063841688d3c961a90a/src/emulator/storage/files.ts#L322-L365) can return nothing in the case of a file not being able to be found - the 404 for this is handled [above](https://github.com/bookcreator/firebase-tools/blob/0191163aa697f2f7cc94c063841688d3c961a90a/src/emulator/storage/apis/gcloud.ts#L274).